File: backend/app/models/meeting.py
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.mixins import IntegerPrimaryKeyMixin, TimestampMixin


class MeetingStatus:
    CREATED = "created"
    UPLOADED = "uploaded"
    TRANSCRIBING = "transcribing"
    TRANSCRIBED = "transcribed"
    ANALYZING = "analyzing"
    COMPLETED = "completed"
    FAILED = "failed"


class Meeting(IntegerPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "meetings"
    __table_args__ = (Index("ix_meetings_created_at", "created_at"),)

    title: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    source_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    recorded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="completed",
        server_default="completed",
        index=True,
    )
    audio_file_path: Mapped[str | None] = mapped_column(String(512), nullable=True)
    audio_filename: Mapped[str | None] = mapped_column(String(255), nullable=True)
    audio_mime_type: Mapped[str | None] = mapped_column(String(100), nullable=True)
    audio_size_bytes: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    participants = relationship(
        "Participant",
        back_populates="meeting",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    transcript_segments = relationship(
        "TranscriptSegment",
        back_populates="meeting",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="TranscriptSegment.sequence_number",
    )
    summary = relationship(
        "Summary",
        back_populates="meeting",
        cascade="all, delete-orphan",
        passive_deletes=True,
        uselist=False,
        single_parent=True,
    )
    action_items = relationship(
        "ActionItem",
        back_populates="meeting",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    chapters = relationship(
        "Chapter",
        back_populates="meeting",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="Chapter.sequence_number",
    )

File: frontend/src/types/meeting.ts
export type MeetingStatus =
  | 'created'
  | 'uploaded'
  | 'transcribing'
  | 'transcribed'
  | 'analyzing'
  | 'completed'
  | 'failed';

export interface Meeting {
  id: number;
  title: string;
  source_name: string | null;
  recorded_at: string;
  duration_ms: number | null;
  status: MeetingStatus;
  audio_file_path: string | null;
  audio_filename: string | null;
  audio_mime_type: string | null;
  audio_size_bytes: number | null;
  workspace_id?: number | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface MeetingListResponse {
  items: Meeting[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface Participant {
  id: number;
  meeting_id: number;
  display_name: string;
  speaker_label: string | null;
  email: string | null;
  is_host: boolean;
}

export interface TranscriptSegment {
  id: number;
  meeting_id: number;
  participant_id: number | null;
  sequence_number: number;
  start_time_ms: number;
  end_time_ms: number;
  text: string;
}

export interface Summary {
  id: number;
  meeting_id: number;
  overview: string;
  key_points: string[];
}

export interface ActionItem {
  id: number;
  meeting_id: number;
  participant_id: number | null;
  description: string;
  is_completed: boolean;
  completed_at: string | null;
  due_at: string | null;
}

export interface Chapter {
  id: number;
  meeting_id: number;
  sequence_number: number;
  title: string;
  summary: string | null;
  start_time_ms: number;
  end_time_ms: number | null;
}

export interface MeetingIntelligence {
  meeting: Meeting;
  summary: Summary | null;
  action_items: ActionItem[];
  chapters: Chapter[];
  participants: Participant[];
  transcript_segments: TranscriptSegment[];
}

export interface MeetingCreateInput {
  title: string;
  source_name?: string;
  recorded_at?: string;
  duration_ms?: number;
  workspace_id?: number;
}
